### ISSUE TITLE
Does torchtune support for traditional text-generation tasks?

### ISSUE THREAD

**Issue by sherlcok314159** [pytorch/torchtune#1249]
I wonder whether torchtune can support traditional tasks such as translation or more general text generation tasks which have a input and output column. I have read the datasets doc at [here](https://pytorch.org/torchtune/stable/api_ref_datasets.html). But it looks like we all need to format it by InstructTemplate. How can I do traditional text-generation tasks by Llama3.1-8B? Any help is welcome. I am new to torchtune.

**Comment by SalmanMohammadi** [comment]
Hey @sherlcok314159. Thanks for your interest in torchtune and for raising this!

It sounds like you're interested in a general next-token-prediction task? Could you possibly provide a link to a dataset you're interested in fine-tuning on?

It sounds like the [text completion dataset](https://docs-preview.pytorch.org/pytorch/torchtune/1147/tutorials/datasets.html#custom-unstructured-text-corpus) could be suitable for your needs, which enables the use of unstructured datasets (and, optionally, the use of sample packing!). 

Let me know if that's pointed you in the right direction - very happy to help more.

**Comment by sherlcok314159** [comment]
Thanks for your quick response. For example, I want to let LLMs to predict the sentiment of movie reviews. https://huggingface.co/datasets/stanfordnlp/sst2. Generally speaking, we can have one or two inputs, and a label which can be a class (classification) or a sequence (generation).

**Comment by SalmanMohammadi** [comment]
Cool!!

So, for classification tasks we support classification models out of the box (e.g. [here's a llama2 model for [reward modelling](https://github.com/pytorch/torchtune/blob/288ff4435b0cf17325b5c3b112f6859a6cdf0ea2/torchtune/models/llama2/_model_builders.py#L279) which can be generalised through the `num_classes` arg). 

You'd then need to add functionality for a classification dataset. @ebsmothers' has far more intelligent things to say than I do: answer [here](https://github.com/pytorch/torchtune/issues/1124#issuecomment-2197037012). To quote:

> I assume you'd need to feed in labels from your dataset in that case (instead of just using shifted tokens as we currently do). 
> If so I think something like our [text_completion_dataset](https://github.com/pytorch/torchtune/blob/88d172fd92237f69b1d18a4693695ca65ca009da/torchtune/datasets/_text_completion.py#L68) could be a good starting point, but we would need to change the labels [here](https://github.com/pytorch/torchtune/blob/88d172fd92237f69b1d18a4693695ca65ca009da/torchtune/datasets/_text_completion.py#L63) to whatever is in your dataset (depending on the format) 
> Finally you would probably want to change [this line](https://github.com/pytorch/torchtune/blob/88d172fd92237f69b1d18a4693695ca65ca009da/recipes/full_finetune_single_device.py#L451) in the training recipe which shifts the labels (because it assumes we are doing next-token prediction). Also cc @RdoubleA here who may have more informed things to say than I do.

Also cc @RdoubleA to register interest in text classification dataset builders/docs (happy to help with this). This has also come up previously in #1124.

**Comment by sherlcok314159** [comment]
Looks like I need to write my own dataset loader script and change the training script to train text classification tasks.

**Comment by SalmanMohammadi** [comment]
I've been working on a reward modelling recipe, so expect this to come to torchtune sometime in the future!

I'll update this weekend with some more guidance and concrete examples to help you get started for now. Let me know how you get on otherwise : )

**Comment by SalmanMohammadi** [comment]
> or a sequence (generation).

And then, I think you should be able to do this out-of-the-box with one of our base models and our [text completion dataset](https://github.com/pytorch/torchtune/blob/288ff4435b0cf17325b5c3b112f6859a6cdf0ea2/torchtune/datasets/_text_completion.py#L75). We also support sample packing here for super speedy training!

**Comment by sherlcok314159** [comment]
Thanks for your kind help! I am trying to use transformers trainer and peft to build a baseline. Then I can compare torchtune and transformers trainer.

**Comment by SalmanMohammadi** [comment]
> Thanks for your kind help! I am trying to use transformers trainer and peft to build a baseline. Then I can compare torchtune and transformers trainer.

Smart! I would've done exactly the same, starting with a baseline helps keeps things sane.

I've been working on RLHF for torchtune, and to train a reward model for it I used [TRL](https://github.com/huggingface/trl). Here's an example of a trained reward model and the command to replicate the training: https://huggingface.co/smohammadi/tinyllama_rm_sentiment_1b.

Note: the trained model can then be used directly in torchtune.

**Comment by rezadnayeri** [comment]
@SalmanMohammadi  super grateful if you could give us some more guidance and examples for the text classification using torchtune. I have been trying unsuccessfully, as I am new with touchtune. It is a fantastic tool, and using it for text classification is super exiting. Can't wait to try it.

**Comment by rezadnayeri** [comment]
any update on this? appreciate your comments.

**Comment by rezadnayeri** [comment]
@RdoubleA any chance you could give us some guidance for the text classification using torchtune. basically, how to prepare the dataset, and how to use mistral_classifier (https://pytorch.org/torchtune/stable/generated/torchtune.models.mistral.mistral_classifier.html#torchtune.models.mistral.mistral_classifier) ?
 thanks a lot!

**Comment by SalmanMohammadi** [comment]
Hey @rezadnayeri. Sorry for the late reply. We're on this now - I'm currently working on adding support for classification datasets out-of-the-box in torchtune alongside @RdoubleA's heroic refactor of our datasets functionality. I'll update very soon with a PR.

**Comment by rezadnayeri** [comment]
Hey @SalmanMohammadi, truly appreciate your effort (and @RdoubleA),  this would be an amazing addition. Meanwhile I will try to be patient :-)

**Comment by qqlabs** [comment]
Hi @SalmanMohammadi, I'm just getting started with torchtune and saw that this classification use case is my exact use case. Wanted to check in on whether you had updates for the classification dataset PR. Thanks!

**Comment by SalmanMohammadi** [comment]
Hey folks. The ball is rolling here. Follow along in https://github.com/pytorch/torchtune/issues/1464! There's a couple TODOs, if you're interested in helping out I'd be happy to guide you through contributing.

**Comment by rezadnayeri** [comment]
@SalmanMohammadi looking forward to this, thank you for your efforts!

**Comment by SalmanMohammadi** [comment]
Would you be able to share any examples of classification datasets you're interested in finetuning on @rezadnayeri @qqlabs @sherlcok314159. The IMDB dataset was raised - it'd really useful to see your use cases, particularly in terms of the format of labels/targets in the dataset.

**Comment by rezadnayeri** [comment]
Hi @SalmanMohammadi : the IMDB dataset is a good benchmark, what I usually do is to follow the same format, "text" is the sequence of string text to be classified, and "label" is the integer ID for the classes (for example 0,1,2,3 for a 4 class problem). Thanks!

**Comment by qqlabs** [comment]
@SalmanMohammadi The IMDB dataset is also a good starting point for me as well, though I am interested in multiclass classification and not just binary. 

My current use case is something like the [Amazon ESCI shopping queries dataset](https://github.com/amazon-science/esci-data). I currently concat the query and product information together into a prompt and am trying to predict one of the classes "exact", "substitute", "complement", and "irrelevant". 

I followed the [tips](https://github.com/pytorch/torchtune/issues/1249#issuecomment-2263197407) from the past comments and modified the current text completion dataset and recipes to do prompt -> completion similar to the [format](https://platform.openai.com/docs/guides/fine-tuning/example-format) for finetuning legacy openai models (babbage-002 and davinci-002). Having some trouble getting the model to actually generate things correctly since I think there are still a lot of next token assumptions that I didn't account for at the generation step - so going to try playing with the reward model to do direct classification next.

**Comment by SalmanMohammadi** [comment]
That's really useful @qqlabs. Thanks so much!

I'm interested in hearing more about how you get on - feel free to ping me on Discord if you have any questions too!

**Comment by rezadnayeri** [comment]
Hello @SalmanMohammadi, wondering if you have any updates on #1464. Thanks
